plugins/in_splunk/splunk_prot.c

```diff
@@ -0,0 +1,822 @@
+            key_str  = (char *) key.via.bin.ptr;
+            key_str_size = key.via.bin.size;
+            check = FLB_TRUE;
+        }
+        if (key.type == MSGPACK_OBJECT_STR) {
+            key_str  = (char *) key.via.str.ptr;
+            key_str_size = key.via.str.size;
+            check = FLB_TRUE;
+        }
+
+        if (check == FLB_TRUE) {
+            if (strncmp(ctx->tag_key, key_str, key_str_size) == 0) {
+                val = (kv+j)->val;
+                if (val.type == MSGPACK_OBJECT_BIN) {
+                    val_str  = (char *) val.via.bin.ptr;
+                    val_str_size = val.via.str.size;
+                    found = FLB_TRUE;
+                    break;
+                }
+                if (val.type == MSGPACK_OBJECT_STR) {
+                    val_str  = (char *) val.via.str.ptr;
+                    val_str_size = val.via.str.size;
+                    found = FLB_TRUE;
+                    break;
+                }
+            }
+        }
+    }
+
+    if (found == FLB_TRUE) {
+        tag = flb_sds_create_len(val_str, val_str_size);
+        if (!tag) {
+            flb_errno();
+            return NULL;
+        }
+        return tag;
+    }
+
+
+    flb_plg_error(ctx->ins, "Could not find tag_key %s in record", ctx->tag_key);
+    return NULL;
+}
+
+/*
+ * Process a raw text payload for Splunk HEC requests, uses the delimited character to split records,
+ * return the number of processed bytes
+ */
+static int process_raw_payload_pack(struct flb_splunk *ctx, flb_sds_t tag, char *buf, size_t size)
+{
+    int ret;
+
+    ret = FLB_EVENT_ENCODER_SUCCESS;
+
+    flb_log_event_encoder_reset(&ctx->log_encoder);
+
+    ret = flb_log_event_encoder_begin_record(&ctx->log_encoder);
+
+    if (ret == FLB_EVENT_ENCODER_SUCCESS) {
+        ret = flb_log_event_encoder_set_current_timestamp(&ctx->log_encoder);
+    }
+
+    if (ret == FLB_EVENT_ENCODER_SUCCESS) {
+        ret = flb_log_event_encoder_append_body_values(
+                &ctx->log_encoder,
+                FLB_LOG_EVENT_CSTRING_VALUE("log"),
+                FLB_LOG_EVENT_STRING_VALUE(buf, size));
+    }
+
+    if (ret == FLB_EVENT_ENCODER_SUCCESS) {
+        ret = flb_log_event_encoder_commit_record(&ctx->log_encoder);
+    }
+
+    if (ret != FLB_EVENT_ENCODER_SUCCESS) {
+        return -1;
+    }
+
+    if (ret == FLB_EVENT_ENCODER_SUCCESS) {
+        flb_input_log_append(ctx->ins, NULL, 0,
+                             ctx->log_encoder.output_buffer,
+                             ctx->log_encoder.output_length);
+    }
+    else {
+        flb_plg_error(ctx->ins, "log event encoding error : %d", ret);
+    }
+
+    return 0;
+}
+
+static int process_json_payload_pack(struct flb_splunk *ctx, flb_sds_t tag, char *buf, size_t size)
+{
+    int ret;
+    size_t off = 0;
+    msgpack_unpacked result;
+    struct flb_time tm;
+    int i = 0;
+    msgpack_object *obj;
+    msgpack_object record;
+    flb_sds_t tag_from_record = NULL;
+
```

Comment:
Actually, we already call it after calling flb_input_log_append.
* https://github.com/fluent/fluent-bit/pull/7349/files#diff-5ba91304a4b9ff74a26cf0d8acbbe30e26c75d18c175bb367cd425bb4ee032e1R323
* https://github.com/fluent/fluent-bit/pull/7349/files#diff-5ba91304a4b9ff74a26cf0d8acbbe30e26c75d18c175bb367cd425bb4ee032e1R387

As far as I know, that resetting function should be called at once before stating the another new record.

Reply:
You do but you don't call it in `process_raw_payload_pack` when there's a failure and the idea is to have a homogeneous behavior which means if you are calling reset before processing a payload for one type then the same behavior should be implemented in the rest.

Even though it's very unlikely that a client sends a malformed raw payload followed by a proper json payload the fact that I noticed that while looking at it means there could be other scenarios (corner cases) where this could happen that I missed (which makes sense).

Additionally, the idea of `flb_log_event_encoder_reset` is that it clears the output buffer and the different inputs so an ideal plugin would : 

1. Reset the encoder
2. Encode N records (where N is a predetermined batch size)
3. Flush the encoder (which involves calling `flb_input_log_append`)

Which means we would be producing reasonably sized record sets for `flb_input_log_append` to place in a chunk (filesystem or memory backed) directly when the plugin is not running in `threaded` mode or more importantly not congesting the ring buffer with a lot of unnecessarily small record set chunks when running in `threaded` mode.

Reply:
Got it. So, there is an important things to handle the msgpack payloads with the same way.
I unified the usage of `flb_log_event_encoder_reset` for raw and hec (batched json) cases.
Both of cases should be cleaned for the succeeded cases.